File: src/ui/Layout.zig
const std = @import("std");
const Surface = @import("../frontend/Surface.zig");
const Editor = @import("../Editor.zig");
const Preview = @import("Preview.zig");
const BrainView = @import("BrainView.zig");
const Self = @This();

pub const Panel = enum {
    file_tree,
    editor,
    preview,
    brain,
};

pub const Rect = struct {
    x: u16,
    y: u16,
    w: u16,
    h: u16,
};

// ── State ─────────────────────────────────────────────────────────────

show_file_tree: bool = true,
show_preview: bool = true,
show_brain: bool = false,
active_panel: Panel = .editor,
// Computed rects
title_rect: Rect = .{ .x = 0, .y = 0, .w = 0, .h = 0 },
tree_rect: Rect = .{ .x = 0, .y = 0, .w = 0, .h = 0 },
editor_rect: Rect = .{ .x = 0, .y = 0, .w = 0, .h = 0 },
preview_rect: Rect = .{ .x = 0, .y = 0, .w = 0, .h = 0 },
brain_rect: Rect = .{ .x = 0, .y = 0, .w = 0, .h = 0 },
status_rect: Rect = .{ .x = 0, .y = 0, .w = 0, .h = 0 },
cmd_rect: Rect = .{ .x = 0, .y = 0, .w = 0, .h = 0 },
width: u16 = 0,
height: u16 = 0,

// ── Layout Calculation ────────────────────────────────────────────────

pub fn compute(self: *Self, w: u16, h: u16) void {
    self.width = w;
    self.height = h;

    // Reserve: 1 row title, 1 row status, 1 row command
    const chrome_rows: u16 = 3;
    const content_h = if (h > chrome_rows) h - chrome_rows else 1;
    const content_y: u16 = 1; // after title

    self.title_rect = .{ .x = 0, .y = 0, .w = w, .h = 1 };
    self.status_rect = .{ .x = 0, .y = h -| 2, .w = w, .h = 1 };
    self.cmd_rect = .{ .x = 0, .y = h -| 1, .w = w, .h = 1 };

    // Panel widths — brain replaces preview (mutually exclusive)
    const tree_w: u16 = if (self.show_file_tree) @min(w / 5, 30) else 0;
    const right_panel = self.show_brain or self.show_preview;
    const right_w: u16 = if (right_panel) @min(w / 4, 40) else 0;
    const editor_w = w -| tree_w -| right_w;

    if (self.show_file_tree) {
        self.tree_rect = .{ .x = 0, .y = content_y, .w = tree_w, .h = content_h };
    }

    self.editor_rect = .{ .x = tree_w, .y = content_y, .w = editor_w, .h = content_h };

    if (self.show_brain) {
        self.brain_rect = .{ .x = tree_w + editor_w, .y = content_y, .w = right_w, .h = content_h };
    } else if (self.show_preview) {
        self.preview_rect = .{ .x = tree_w + editor_w, .y = content_y, .w = right_w, .h = content_h };
    }
}

pub fn togglePanel(self: *Self, panel: Panel) void {
    switch (panel) {
        .file_tree => self.show_file_tree = !self.show_file_tree,
        .preview => {
            self.show_preview = !self.show_preview;
            if (self.show_preview) self.show_brain = false; // mutually exclusive
        },
        .brain => {
            self.show_brain = !self.show_brain;
            if (self.show_brain) self.show_preview = false; // mutually exclusive
        },
        .editor => {},
    }
}

pub fn cyclePanel(self: *Self) void {
    self.active_panel = switch (self.active_panel) {
        .file_tree => .editor,
        .editor => if (self.show_brain) .brain else if (self.show_preview) .preview else if (self.show_file_tree) .file_tree else .editor,
        .preview => if (self.show_file_tree) .file_tree else .editor,
        .brain => if (self.show_file_tree) .file_tree else .editor,
    };
}

// ── Rendering ─────────────────────────────────────────────────────────

pub fn renderChrome(self: *Self, surface: *Surface) void {
    const tc = @import("../themes.zig").currentColors();
    // Title bar
    surface.fillRow(self.title_rect.y, ' ', tc.title_fg, tc.title_bg, .{});
    const title = " lazy-md v0.1.0";
    surface.putStr(0, 0, title, tc.title_fg, tc.title_bg, .{ .bold = true });

    // Keyboard hints on title bar (right-aligned)
    const hints = "Tab:panels  1:tree  2:preview  3:brain  :q quit ";
    if (hints.len < self.width) {
        surface.putStr(self.width -| @as(u16, @intCast(hints.len)), 0, hints, tc.border_active, tc.title_bg, .{});
    }

    // Panel borders
    if (self.show_file_tree and self.tree_rect.w > 0) {
        surface.drawVLine(self.tree_rect.x + self.tree_rect.w -| 1, self.tree_rect.y, self.tree_rect.h, tc.border, .default);
    }
    if (self.show_brain and self.brain_rect.w > 0) {
        surface.drawVLine(self.brain_rect.x, self.brain_rect.y, self.brain_rect.h, tc.border, .default);
    } else if (self.show_preview and self.preview_rect.w > 0) {
        surface.drawVLine(self.preview_rect.x, self.preview_rect.y, self.preview_rect.h, tc.border, .default);
    }
}

pub fn renderFileTree(self: *Self, surface: *Surface, entries: []const FileEntry) void {
    if (!self.show_file_tree) return;

    const r = self.tree_rect;
    const tc = @import("../themes.zig").currentColors();
    const is_active = self.active_panel == .file_tree;
    const border_fg: Surface.Color = if (is_active) tc.border_active else tc.border;

    // Panel header
    surface.putStr(r.x + 1, r.y, " Files ", tc.title_fg, .default, .{ .bold = true });
    surface.drawVLine(r.x + r.w -| 1, r.y, r.h, border_fg, .default);

    // File entries
    const max_entries = if (r.h > 2) r.h - 1 else 0;
    for (entries, 0..) |entry, i| {
        if (i >= max_entries) break;
        const y = r.y + 1 + @as(u16, @intCast(i));
        const icon: []const u8 = if (entry.is_dir) "  " else "  ";
        const fg: Surface.Color = if (entry.is_dir) .bright_blue else if (entry.is_md) .bright_green else .white;
        surface.putStrTrunc(r.x + 1, y, icon, r.w -| 2, fg, .default, .{});
        surface.putStrTrunc(r.x + 3, y, entry.name, r.w -| 4, fg, .default, .{});
    }
}

pub fn renderPreview(self: *Self, surface: *Surface, editor: *Editor, preview: *Preview) void {
    if (!self.show_preview) return;

    const r = self.preview_rect;
    const tc = @import("../themes.zig").currentColors();
    const is_active = self.active_panel == .preview;
    const border_fg: Surface.Color = if (is_active) tc.border_active else tc.border;

    // Panel border and header
    surface.drawVLine(r.x, r.y, r.h, border_fg, .default);
    surface.putStr(r.x + 1, r.y, " Preview ", tc.title_fg, .default, .{ .bold = true });

    // Rendered markdown preview
    preview.render(surface, editor, r);
}

pub fn renderBrain(self: *Self, surface: *Surface, brain: *BrainView) void {
    if (!self.show_brain) return;

    const r = self.brain_rect;
    const tc = @import("../themes.zig").currentColors();
    const is_active = self.active_panel == .brain;
    const border_fg: Surface.Color = if (is_active) tc.border_active else tc.border;

    // Panel border and header
    surface.drawVLine(r.x, r.y, r.h, border_fg, .default);
    surface.putStr(r.x + 1, r.y, " Brain ", tc.title_fg, .default, .{ .bold = true });

    // Render graph inside panel (offset by 1 for border, 1 for header)
    const inner: Rect = .{
        .x = r.x + 1,
        .y = r.y + 1,
        .w = if (r.w > 2) r.w - 2 else 1,
        .h = if (r.h > 2) r.h - 2 else 1,
    };
    brain.render(surface, inner);
}

pub const FileEntry = struct {
    name: []const u8,
    is_dir: bool,
    is_md: bool,
};

// ── Tests ─────────────────────────────────────────────────────────────

test "layout computation" {
    var layout = Self{};
    layout.compute(120, 40);

    try std.testing.expect(layout.editor_rect.w > 0);
    try std.testing.expect(layout.tree_rect.w > 0);
    try std.testing.expect(layout.preview_rect.w > 0);
    try std.testing.expectEqual(@as(u16, 0), layout.title_rect.y);
}

test "toggle panels" {
    var layout = Self{};
    layout.compute(120, 40);

    const old_tree_w = layout.tree_rect.w;
    layout.togglePanel(.file_tree);
    layout.compute(120, 40);
    try std.testing.expect(!layout.show_file_tree);
    _ = old_tree_w;
}

File: internal/ui/layout.go
package ui

// Panel identifies a UI panel.
type Panel int

const (
	PanelEditor Panel = iota
	PanelFileTree
	PanelPreview
	PanelBrain
)

// Rect holds computed dimensions for a panel.
type Rect struct {
	W, H int
}

// LayoutState manages panel layout and visibility.
type LayoutState struct {
	ActivePanel  Panel
	ShowFileTree bool
	ShowPreview  bool
	ShowBrain    bool

	TreeRect    Rect
	EditorRect  Rect
	PreviewRect Rect
	BrainRect   Rect

	// Status/command bar heights
	TitleH   int
	StatusH  int
	CommandH int
}

// NewLayout creates a default layout state.
func NewLayout() LayoutState {
	return LayoutState{
		ActivePanel:  PanelEditor,
		ShowFileTree: true,
		ShowPreview:  true,
		ShowBrain:    false,
		TitleH:       1,
		StatusH:      1,
		CommandH:     1,
	}
}

// Compute recalculates panel dimensions based on terminal size.
func (l *LayoutState) Compute(w, h int) {
	bodyH := h - l.TitleH - l.StatusH - l.CommandH
	if bodyH < 1 {
		bodyH = 1
	}

	treeW := 0
	if l.ShowFileTree {
		treeW = w / 5
		if treeW > 30 {
			treeW = 30
		}
		if treeW < 10 {
			treeW = 10
		}
	}

	rightW := 0
	if l.ShowPreview || l.ShowBrain {
		rightW = w / 3
		if rightW > 50 {
			rightW = 50
		}
		if rightW < 15 {
			rightW = 15
		}
	}

	editorW := w - treeW - rightW
	if editorW < 10 {
		editorW = 10
	}

	l.TreeRect = Rect{W: treeW, H: bodyH}
	l.EditorRect = Rect{W: editorW, H: bodyH}

	if l.ShowBrain {
		l.BrainRect = Rect{W: rightW, H: bodyH}
		l.PreviewRect = Rect{}
	} else if l.ShowPreview {
		l.PreviewRect = Rect{W: rightW, H: bodyH}
		l.BrainRect = Rect{}
	} else {
		l.PreviewRect = Rect{}
		l.BrainRect = Rect{}
	}
}

// TogglePanel toggles a panel's visibility.
func (l *LayoutState) TogglePanel(p Panel) {
	switch p {
	case PanelFileTree:
		l.ShowFileTree = !l.ShowFileTree
		if !l.ShowFileTree && l.ActivePanel == PanelFileTree {
			l.ActivePanel = PanelEditor
		}
	case PanelPreview:
		l.ShowPreview = !l.ShowPreview
		if l.ShowPreview {
			l.ShowBrain = false
		}
		if !l.ShowPreview && l.ActivePanel == PanelPreview {
			l.ActivePanel = PanelEditor
		}
	case PanelBrain:
		l.ShowBrain = !l.ShowBrain
		if l.ShowBrain {
			l.ShowPreview = false
		}
		if !l.ShowBrain && l.ActivePanel == PanelBrain {
			l.ActivePanel = PanelEditor
		}
	}
}

// CyclePanel moves focus to the next visible panel.
func (l *LayoutState) CyclePanel() {
	// Build ordered list: FileTree → Editor → Preview/Brain
	var panels []Panel
	if l.ShowFileTree {
		panels = append(panels, PanelFileTree)
	}
	panels = append(panels, PanelEditor)
	if l.ShowPreview {
		panels = append(panels, PanelPreview)
	}
	if l.ShowBrain {
		panels = append(panels, PanelBrain)
	}

	for i, p := range panels {
		if p == l.ActivePanel {
			l.ActivePanel = panels[(i+1)%len(panels)]
			return
		}
	}
	l.ActivePanel = PanelEditor
}

// HitTest returns which panel occupies a given x coordinate.
func (l *LayoutState) HitTest(x, _ int) Panel {
	offset := 0
	if l.ShowFileTree {
		offset += l.TreeRect.W
		if x < offset {
			return PanelFileTree
		}
	}
	offset += l.EditorRect.W
	if x < offset {
		return PanelEditor
	}
	if l.ShowBrain {
		return PanelBrain
	}
	if l.ShowPreview {
		return PanelPreview
	}
	return PanelEditor
}
